### ISSUE TITLE
Voiceover does not place cursor into dialog windows (VPAT #13)

### ISSUE THREAD

**Issue by abaevbog** [zotero/zotero#3936]
When voiceover is active, opening any dialog window (e.g "New collection") does not place voiceover's cursor inside of the new window. It remains within the original window even though the focus has moved into the dialog. So far, no focus tweaks or adding `role=dialog`/`aria-modal="true"` helped. In the screenshot below, the focus is inside of the "New collection" window but the small black square in the top left corner is the voiceover's cursor. Interestingly, opening a window not as a `modal` doesn't cause such issue but it does looks different and breaks the current setup of how dialogs return selected values.

<img width="1415" alt="Screenshot 2024-04-08 at 6 09 55 PM" src="https://github.com/zotero/zotero/assets/36271954/883c4bfa-964b-4cf8-91ca-bb086b0fa6ff">

A possible (though unfortunate) solution would be to not open the dialogs as a separate window but display them within the main `ZoteroPane` window with an opaque background covering the rest of the screen. Something similar to what is currently done in `Zotero.showZoteroPaneProgressMeter` with `zotero-pane-overlay`. This seems to be what [Firefox does](https://searchfox.org/mozilla-central/source/browser/base/content/main-popupset.inc.xhtml#238) as well, for example in `about:preferences#general` > "Network settings".
That way, everything happens within the same window, and voiceover just follows the focused element. 

The main drawback, of course, is that it requires revising all the dialog windows.

**Comment by abaevbog** [comment]
> The main drawback, of course, is that it requires revising all the dialog windows.

To avoid redoing the logic of displaying pretty much every single `<dialog>` window, maybe we can:
1. make dialog windows non-modal
2. pass a callback function into the dialog window that will run on accept
3. cover the screen with an opaque element to make it look at least somewhat similar to what it was before.

In this case, the dialog window can be be moved around and it might open in a slightly different location than now but besides that, it should function pretty much the same way. I have a proof of concept here of what it could look like: https://github.com/abaevbog/zotero/commit/e0c421542ed497824581e8fc0f6fa9d001f8fc2b




https://github.com/zotero/zotero/assets/36271954/1d70be25-89c9-4ac8-a936-1fd5a115c374


I think that restructuring the `dialogs` to exist within our main window would be the least hacky solution but it is likely more tricky and time consuming, since it would involve fundamentally changing how all dialogs are handled. This is a quicker fix to handle voiceover issue for vpat. @dstillman which of these directions do you think is better? Or maybe I am missing something obvious and it's none of these?

**Comment by adomasven** [comment]
Is this an issue in Thunderbird or if opening some of the dialogs that still have a separate window in Firefox (e.g. the security certificate window)?

**Comment by abaevbog** [comment]
> Is this an issue in Thunderbird

It is. For example, when the "Chat" section is selected in the left sidebar and the chat account is not yet setup, clicking the "Get started" button will open up a modal window. In a similar fashion, voiceover keeps the cursor where it was on the main window. 

> or if opening some of the dialogs that still have a separate window in Firefox (e.g. the security certificate window)?

On firefox, there are few windows opened as a modal on macOS. Security certificate is a non-modal (I can drag it around). In a few places, windows are opened as modals on win/linux and as non-modal specifically on mac OS. For example, [set as a background image](https://searchfox.org/mozilla-central/source/browser/base/content/nsContextMenu.js#1825) window that can be called from context menu on an image. Not sure if this is relevant for this specific issue.

I did run into the [reset profile modal](https://searchfox.org/mozilla-central/source/browser/base/content/safeMode.js#28) (`about:support` > "Refresh firefox") that is opened the same way via `window.openDialog` and `modal` is one of the options. The difference there is that it creates a `<browser>` inside a <`html:dialog id="window-modal-dialog">` component at the bottom of the page. There is some logic of what is going on [here](https://searchfox.org/mozilla-central/source/browser/base/content/browser.js#9010), so I'll try to see if we can recreate the same behavior.

**Comment by abaevbog** [comment]
I'm digging into the source code to get some more insight into this. But one more observation is that `window.openDialog` seems ... different when it is called in the reset profile modal vs if called directly from the console via browser toolbox. If I press the "Refresh firefox" button, it acts as described above and voiceover is happy with it. If I just run the same 
`window.openDialog( "chrome://global/content/resetProfile.xhtml", null, "chrome,modal,centerscreen,titlebar,dialog=yes", {})` from the console, I get the same behavior as with the "New collection" dialog.

**Comment by abaevbog** [comment]
I have another idea that is inspired by what is happening with "Reset Firefox" - use a wrapper dialog with a `<browser>` component inside of it that will render the `.xhtml` files. Then, we can display the wrapper dialog with `.showDialog()` and pass data from and into the scripts from `.xhtml` files through the `<browser>`.  The benefit here is that the window remains an actual modal without completely redoing how dialogs are handled with the`.xhtml` templates and their scripts. It does require a bit of refactoring to pass data back and forth (since we can't pass the `io` parameter as with `window.openDialog`) though. I have a proof of concept specifically for the "New collection" dialog here https://github.com/abaevbog/zotero/commit/122a7cc4981507bbc95cca0de2d92e05eff35564

This is roughly what it looks like (styling/sizing is off as it's just proof of concept). Note that the black square moves inside of the dialog and then goes back onto the "New collection" button:

https://github.com/zotero/zotero/assets/36271954/d2607a19-3abf-4edc-b9d8-5ce72ac898ea

**Comment by adomasven** [comment]
Until some additional testing here I didn't realize how much more different the macOS handling of modal dialogs is compared to Linux (and probably Windows). Newer versions of Firefox are actually largely getting rid of popup windows completely and display all dialogs in this kind of fake modal way, and I wonder if we're going to get something that works out of the box with fx128. Otherwise, this solution seems decent, and probably wouldn't even need to handle this separately on Linux and Windows. This behaviour for modal dialogs is more inline with what the intention of a modal dialog is anyway.

**Comment by dstillman** [comment]
Thanks for exploring this, @abaevbog. I would really prefer not use fake modals — it's one of the biggest tells that something is a web-based, non-native app. (And we definitely don't want to use non-modals where modals are appropriate.)

Firefox has been switching to fake modals for years so that alerts in one browser tab don't block browsing in other tabs, though it's true that they now appear to be doing it for non-tab-specific modals as well (e.g., Help → Troubleshooting Mode…).

Have we tested whether this still happens in current Firefox versions? It seems like we should just report this to Firefox as an accessibility, though they may not be eager to fix it if they're getting rid of all real modal windows — we'd probably need to find at least one example of a real modal. (Are nsIPromptService windows becoming fake modals? If so, I guess this is somewhat moot.)

**Comment by dstillman** [comment]
(We could also try to actually patch it in Firefox, though that might be a tall order.)

**Comment by adomasven** [comment]
TB is not necessarily going the same way regarding modals and since they are also affected, maybe reporting this indeed is worthwhile.

**Comment by abaevbog** [comment]
> Have we tested whether this still happens in current Firefox versions?

I was examining this on firefox `124.0.2`, so all of my firefox-specific examples/notes are from that. I assumed this could be something in the process of being changed so I wanted to check on the latest version available. 

Reporting it would make sense. If it helps, I finally ran into this exact issue within firefox itself. "about:support" > "Show Update History" opens the modal in the same way and if you have voiceover turned on, you'll notice that the cursor remains on the toolbar of the main window instead of going inside of the dialog, even though the focus is inside of it. Then, voiceover will say that you are on the "tabs toolbar item pallete" which is not right.

<img width="978" alt="Screenshot 2024-04-10 at 9 35 36 AM" src="https://github.com/zotero/zotero/assets/36271954/70d2cb79-ff3d-42f6-9b8d-f0a30b8b9795">

**Comment by abaevbog** [comment]
This might be an obvious observation but I noticed that `Services.focus.activeWindow` and `Services.focus.focusedWindow` are different on windows vs macOS when the dialogs are opened. For the `newCollection` dialog, `zoteroPane.xhtml` is the active and focused window, while on windows it's `newCollection.xhtml`. This might be somewhat relevant here because voiceover could think that `zoteroPane` is the main window, so it keeps the cursor in it.

**Comment by adomasven** [comment]
Not sure if you bothered, but there's some focus related stuff going on in fx `dialog.js` https://searchfox.org/mozilla-central/source/toolkit/content/widgets/dialog.js#332

**Comment by dstillman** [comment]
So this was fixed by https://bugzilla.mozilla.org/show_bug.cgi?id=1884631, which replaced sheets with modal dialogs. Available now in the Firefox 126 beta.

I've backported the change, and it works. Unfortunately it's much uglier than before, and also more confusing, since it no longer dims the background:

![Screenshot 2024-04-18 at 5 48 45 AM](https://github.com/zotero/zotero/assets/783082/815d55fc-209e-435c-864a-ff2897c3e933)

Before:

![Screenshot 2024-04-18 at 5 50 15 AM](https://github.com/zotero/zotero/assets/783082/334b7c8e-cee2-4dce-aef1-4be61251f806)

This somewhat defeats the purpose of using native dialogs vs. fake dialogs of our own (or Mozilla's) design, so we maybe will need to revisit that. But I guess we have to go with this for now.

**Comment by dstillman** [comment]
Oh, and for `nsIPromptService` dialogs it now shows the title twice, once in the title bar and once in the window. Maybe we can fix that in commonDialog.xhtml?

**Comment by dstillman** [comment]
Unless we can just make all of these title-bar-less windows? @windingwind, do we have an easy way to do that now?

**Comment by windingwind** [comment]
It works (though without the fade-in/out animation and grey-out mask of the background window, but that's relatively less important, and we can fake one):

<img width="419" alt="image" src="https://github.com/zotero/zotero/assets/33902321/deecc3be-146c-479f-bbd5-bb64e6495893">

<img width="617" alt="image" src="https://github.com/zotero/zotero/assets/33902321/e7458a82-b8c6-44e9-a28a-fb6a91fbc9fe">

**Comment by windingwind** [comment]
[#4011](https://github.com/zotero/zotero/pull/4011)

**Comment by dstillman** [comment]
Awesome!
